src/Appwrite/Auth/OAuth2.php

```diff
@@ -62,9 +62,9 @@ abstract public function getLoginURL():string;
     /**
      * @param string $code
      *
-     * @return string
+     * @return array
      */
-    abstract public function getAccessToken(string $code):string;
+    abstract public function getTokens(string $code):array;
```

Comment:
A better developer friendly signature might be to have both `getAccessToken` and `getRefreshToken` methods where each returns a string. We can store the result of the first API call and cache it in `private $tokens` if `$this->tokens` is not empty we can skip a second API call and just return the relevant string according to the method needs.

We follow a similar pattern for `getUserEmail`, `getUserID`, and `getUserName`.

Reply:
User's actually have `getUser` that gets the data, and functions `getUserEmail` `getUserName`to return specific ones.

To follow that, we would have `getTokens`, store the response in variable, and then return specific parts with functions `getAccessToken` and `getRefreshToken`.

Hope that is what you meant. I have implemented in Sporify, please let me know what you think, and I can add it elsewhere: https://github.com/appwrite/appwrite/pull/2733/files#diff-00ac4df20657d38a00b1d590988c2daee3ba5476f65b9b915ef3851cbc453deb

(getRefreshToken, getTokenExpiry and getAccessToken will be in abstract class to prevent repetition)

Reply:
Yes, this example look good, but `getTokens` can be a private method.

Reply:
Changed it to 'protected'. I added some universal functions to OAuth class, and it now needs to access this `getTokens` method.